### ISSUE TITLE
Failing to create Vulkan instance on Ubuntu 18.04 - "libGLESv1_CM.so" cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by mitchmindtree** [baldurk/renderdoc#1279]
## Description

Attempting to launch an application using qrenderdoc with a Vulkan graphics instance on Ubuntu 18.04.01 LTS results in the following output:

```
thread 'main' panicked at 'failed to create vulkan instance: LoadingError(LibraryLoadFailure("libGLESv1_CM.so: cannot open shared object file: No such file or directory"))'
```

The application I'm attempting to launch uses the Rust Vulkan API called [vulkano](https://github.com/vulkano-rs/vulkano). The application `panic!`s when attempting to create a Vulkan instance seemingly due to the error mentioned above.

Fwiw, I'm able to run the application itself fine without qrenderdoc. I'm also able to run the application with qrenderdoc on my Arch Linux machine seemingly without issues. It seems as though perhaps this `libGLESv1_CM.so` library is not where it is expected to be?

Here is the output of `vulkaninfo`:

[vulkaninfo.txt](https://github.com/baldurk/renderdoc/files/2896614/vulkaninfo.txt)

## Repro steps

I don't have the exact steps I took on hand, but the steps went something like this:

- Install Ubuntu 18.04
- Update and upgrade system
- Install nvidia vulkan drivers
- Install renderdoc dependencies (I checked that those mentioned in the README were installed, don't remember if they were already installed by default with Ubuntu or not).
- Run `qrenderdoc` from Terminal. Follow the vulkan setup notification prompts. Install for system-wide use by elevating permissions.
- Launch application from within `qrenderdoc`.

## Environment

- RenderDoc build: 2019-02-21 nightly build. (also tried with current stable with same results).
- Operating System: Ubuntu 18.04.01 LTS
- API: Vulkan
- GPU: Nvidia 1060 (see vulkaninfo output)

cc @freesig @joshuabatty

**Comment by baldurk** [comment]
Do you have a callstack for what code is actually trying to load `libGLESv1_CM.so`? That library is only used for GLES and has nothing to do with vulkan. I quickly searched the vulkano codebase and it doesn't seem to be used there.

RenderDoc will hook it if present since it doesn't know in advance if the application will be using vulkan or GL, but that only involves loading it if it's available - if it's not available then it won't be loaded.

**Comment by mitchmindtree** [comment]
Thanks a lot for the brief response, much appreciated!

Here's the stack trace:

```
Log     - Got remote handshake: microbe_mirror [5897]                                                  
thread 'main' panicked at 'failed to create vulkan instance: LoadingError(LibraryLoadFailure("libGLESv1_CM.so: cannot open shared object file: No such file or directory"))', src/libcore/result.rs:1009:5
stack backtrace:                                                                
   0:     0x55dee8a263a3 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h00d1e05a61bd440b
                               at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                       
   1:     0x55dee8a208a8 - std::sys_common::backtrace::_print::hc0d53aca8da62f75
                               at src/libstd/sys_common/backtrace.rs:71                                    
   2:     0x55dee8a25142 - std::panicking::default_hook::{{closure}}::h46d30bcc4bfff149                           
                               at src/libstd/sys_common/backtrace.rs:59                                    
                               at src/libstd/panicking.rs:211                                                     
   3:     0x55dee8a24ead - std::panicking::default_hook::h017696c2a8b7b16f
                               at src/libstd/panicking.rs:227                   
   4:     0x55dee8a25850 - std::panicking::rust_panic_with_hook::h8cbdfe43764887be
                               at src/libstd/panicking.rs:491                                        
   5:     0x55dee8a253d1 - std::panicking::continue_panic_fmt::h3d3c5a833c00a5e1
                               at src/libstd/panicking.rs:398                                                                                                                                             
   6:     0x55dee8a252b5 - rust_begin_unwind                 
                               at src/libstd/panicking.rs:325                                          
   7:     0x55dee8a3bddc - core::panicking::panic_fmt::h4d67173bc68f6d5a                                  
                               at src/libcore/panicking.rs:95                   
   8:     0x55dee7e19220 - core::result::unwrap_failed::hf76f535430b160f1                                  
                               at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libcore/macros.rs:26        
   9:     0x55dee7e0ddaf - <core::result::Result<T, E>>::expect::h769e342efb751ee6                         
                               at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libcore/result.rs:835       
  10:     0x55dee7882927 - <nannou::app::Builder<M, E>>::run::{{closure}}::hb66c11cfe6040020
                               at /home/dms-gf/.cargo/git/checkouts/nannou-d8ba0e7df724bf6f/8267454/src/app.rs:573
  11:     0x55dee78ceef7 - <core::option::Option<T>>::unwrap_or_else::h4220e46ed52ec2d4
                               at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libcore/option.rs:396
  12:     0x55dee7881819 - <nannou::app::Builder<M, E>>::run::h3a6a54c5e5c679a6                                                                                                                           
                               at /home/dms-gf/.cargo/git/checkouts/nannou-d8ba0e7df724bf6f/8267454/src/app.rs:547
  13:     0x55dee7872c8f - microbe_mirror::main::h1e9e7b7c67d306aa                                     
                               at src/bin/main.rs:10                               
  14:     0x55dee78c65ff - std::rt::lang_start::{{closure}}::h1619699e28e487e9  
                               at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libstd/rt.rs:74
  15:     0x55dee8a25252 - std::panicking::try::do_call::h69790245ac2d03fe             
                               at src/libstd/rt.rs:59                                                     
                               at src/libstd/panicking.rs:310                     
  16:     0x55dee8a35899 - __rust_maybe_catch_panic                                                        
                               at src/libpanic_unwind/lib.rs:102                            
  17:     0x55dee8a25ce3 - std::rt::lang_start_internal::h540c897fe52ba9c5                                        
                               at src/libstd/panicking.rs:289                          
                               at src/libstd/panic.rs:398                                                  
                               at src/libstd/rt.rs:58                                                                                                                                                     
  18:     0x55dee78c65d8 - std::rt::lang_start::h48d95f802e320792                                                 
                               at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libstd/rt.rs:74  
  19:     0x55dee7874c59 - main                                                    
  20:     0x7fee923b8b96 - __libc_start_main                                    
  21:     0x55dee786db19 - _start                                                                    
  22:                0x0 - <unknown>                                                                      
```

@freesig mentions that this stacktrace might be a bit misleading and that he has some more useful info - he'll post it here soon.

**Comment by baldurk** [comment]
FYI I would definitely need more information to investigate further. I don't know anything about rust really, but I tried to look up the code and either those line numbers don't match up or I'm looking in the wrong place because I couldn't find any code referencing `libGLESv1_CM.so`.

**Comment by baldurk** [comment]
I also tried running the sample in the README from [nannou](https://github.com/nannou-org/nannou) but it worked OK in my tests, even if I remove the GLESv1 library from my system search paths.

**Comment by freesig** [comment]
Hey sorry for the delay on replying. 
If you tried nannou from master it's still on OpenGL. All our Vulkano support is on the `v0.9` branch.

[This](https://github.com/vulkano-rs/vulkano/blob/912abdf40f58123d618bb187159da3f70cd7e3fe/vulkano/src/instance/loader.rs#L189) is the code in vulkano that gets called. It calls `dlopen(libvulkan.so.1)` under the hood. I think renderdoc is perhaps swapping this out for `libGLESv1_CM.so`.

[This](https://github.com/tomaka/shared_library) is the rust library that `vulkano` is using under the hood to dynamically load the vulkan library

**Comment by baldurk** [comment]
No RenderDoc doesn't change a call to open `libvulkan.so.1` to `libGLESv1_CM.so`, that wouldn't make sense since there's no expectation that GLES will be available and they have different entry points.

On the `v0.9` branch I was able to repro it. I still don't understand why the backtrace is pointing to completely different code but I don't know rust.

It seems like the problem is the application is calling `dlerror()` and then fatal erroring if anything is returned, even if the `dlopen` worked perfectly well. So what it sees is after calling `dlopen(libvulkan.so.1)` RenderDoc processes all its libraries it's going to hook to see if they're present with `RTLD_NOLOAD`. For some reason even this call if it fails will set the `dlerror()` which then causes the program to fatal error. I can change my code to `libfoobarblah.so` and that will show up in the error message.

I consider it an application bug (albeit a minor one since without renderdoc interfering it's unlikely it would ever fire), but I've added a call myself in to `dlerror()` in that commit above after my processing just to clear any false positive error flags that might set off sensitive code.
